File: src/main/java/com/ethgasviewer/server/model/HarvestDTO.java
package com.ethgasviewer.server.model;

import lombok.Data;

import javax.persistence.Cacheable;
import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.Index;
import javax.persistence.Table;
import java.math.BigInteger;

@Entity
@Table(name = "harvest_tx", indexes = {
    @Index(name = "idx_harvest_tx", columnList = "blockDate"),
    @Index(name = "idx_harvest_tx2", columnList = "methodName, vault")
})
@Cacheable(false)
@Data
public class HarvestDTO{

    @Id
    private String id;
    private String hash;
    private BigInteger block;
    private boolean confirmed = false;
    private Long blockDate;
    private String methodName;
    private String owner;
    private Double amount;
    private Double amountIn;
    private String vault;
    private Double lastGas;
    private Double lastTvl;
    private Double lastUsdTvl;
    private Integer ownerCount;
    private Double sharePrice;
    private Long usdAmount;
    private String prices;
    private String lpStat;
}

File: src/app/models/harvest-dto.ts
import {PricesDto} from './prices-dto';
import {LpStat} from './lp-stat';

export class HarvestDto {
  id: string;
  hash: string;
  block: number;
  confirmed: boolean;
  blockDate: number;
  network: string;
  methodName: string;
  owner: string;
  amount: number;
  amountIn: number;
  vault: string;
  lastGas: number;
  lastTvl: number;
  lastUsdTvl: number;
  ownerCount: number;
  sharePrice: number;
  usdAmount: number;
  lpStat: string;
  ownerBalance: number;
  ownerBalanceUsd: number;
  allOwnersCount: number;
  allPoolsOwnersCount: number;
  underlyingPrice: number;
  profit: number;
  profitUsd: number;
  totalAmount: number;

  lpStatDto: LpStat;
  blockDateAdopted: Date;
  acquired: Date;

  public static fromJson(data: string): HarvestDto {
    const tx: HarvestDto = new HarvestDto();
    Object.assign(tx, JSON.parse(data));
    tx.acquired = new Date();
    HarvestDto.enrich(tx);
    return tx;
  }

  public static enrich(tx: HarvestDto): void {
    if (tx.acquired == null) {
      tx.acquired = new Date();
    }
    if (tx.lpStat) {
      tx.lpStatDto = JSON.parse(tx.lpStat);
    }
    HarvestDto.fillBlockDateAdopted(tx);
  }

  public static fillBlockDateAdopted(tx: HarvestDto): void {
    if (tx.blockDateAdopted == null) {
      const d = new Date(0);
      d.setUTCSeconds(tx.blockDate);
      tx.blockDateAdopted = d;
    }
  }

  print(): string {
    // (moment(this.blockDateAdopted)).format('DD-MMM HH:mm:ss')
    return this.methodName
        + ' ' + this.usdAmount?.toFixed(2)
        + '$ ' + this.vault
        + ' TVL is ' + this.lastUsdTvl?.toFixed(2) + '$'
        ;
  }
}
